In [1]:
import pandas as pd
import numpy as np

# HECHOS - HOMICIDIOS

In [9]:
df = pd.read_excel('../Data/excel/homicidios.xlsx',['HECHOS','VICTIMAS'])

In [10]:
df_hechos = df['HECHOS']

In [11]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

In [4]:
df_hechos.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

se observan muy pocas columnas con campos en null, procedemos a eliminar columnas donde se hace rebundante la info y renombrar las que necesitamos

In [12]:
columnas = ['AAAA', 'MM', 'DD','HH','PARTICIPANTES','Calle','Cruce','Dirección Normalizada']
df_hechos = df_hechos.drop(columnas, axis=1)

In [13]:
renombrar = {'ID': 'Id_hechos',
             'N_VICTIMAS': 'nro_victimas',
             'FECHA': 'fecha',
             'HORA': 'hora',
             'LUGAR_DEL_HECHO': 'lugar_hecho',
             'TIPO_DE_CALLE': 'tipo_calle',
             'Altura': 'altura',
             'COMUNA': 'comuna',
             'XY (CABA)': 'geocodificacion_CABA',
             'pos x': 'longitud',
             'pos y	': 'latutid',
             'VICTIMA': 'vehiculo_victima',
             'ACUSADO': 'vehiculo_acusado'}

df_hechos = df_hechos.rename(columns=renombrar)
df_hechos = df_hechos.rename(columns={'pos y': 'latitud'})

corremos una funcion a nuestradata para que sea mas legible

In [15]:
def transformar_texto(columna):
    if pd.notnull(columna):
        return columna.strip().lower().capitalize().lower().capitalize()
    else:
        return columna 

def transformar_dataframe(df, columnas_a_transformar):
    for columna in columnas_a_transformar:
        df[columna] = df[columna].apply(transformar_texto)

In [16]:
columnas = ['lugar_hecho', 'tipo_calle','vehiculo_victima','vehiculo_acusado']
transformar_dataframe(df_hechos, columnas)

In [24]:
df_hechos.sample(5)

Id_hechos  nro_victimas      fecha      hora  \
117  2016-0148             1 2016-11-16  22:20:00   
69   2016-0093             1 2016-07-17  03:30:00   
23   2016-0034             1 2016-03-12  12:30:00   
253  2017-0137             1 2017-11-17  11:45:00   
110  2016-0140             1 2016-10-31  02:58:00   

                                 lugar_hecho tipo_calle  altura  comuna  \
117                        Av garmendia 4895    Avenida  4895.0      15   
69          Av juan b alberdi y emilio mitre    Avenida     NaN       6   
23             Av amancio alcorta y bonavena    Avenida     NaN       4   
253  Av. gral. enrique mosconi y san nicolas    Avenida     NaN      11   
110        Av juramento y vuelta de obligado    Avenida     NaN      13   

                        geocodificacion_CABA      longitud       latitud  \
117   Point (99810.88349598 103818.87868694)  -58.46536113  -34.59484504   
69   Point (101804.61714036 100693.45218194)  -58.44362126  -34.62301748   
23    Point (105230.98315750 97613.97258373)  -58.40623949  -34.65076549   
253   Point (95458.27090734 104045.28333021)  -58.51280715  -34.59279388   
110  Point (100700.71684791 107520.92282315)  -58.45566399  -34.56147293   

    vehiculo_victima vehiculo_acusado  
117             Moto             Moto  
69              Auto             Auto  
23            Peaton        Pasajeros  
253           Peaton           Cargas  
110        Pasajeros        Pasajeros

recordamos que tenemos 567 valores como nulos en nuestra columna altura,  procedemos a castearla para dejar en tipo entero

In [43]:
df_hechos.isnull().sum()

Id_hechos                 0
nro_victimas              0
fecha                     0
hora                      0
lugar_hecho               0
tipo_calle                0
altura                  567
comuna                    0
geocodificacion_CABA      0
longitud                  0
latitud                   0
vehiculo_victima          0
vehiculo_acusado          0
dtype: int64

# VICTIMAS

In [28]:
df_victimas = df['VICTIMAS']

In [16]:
df_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


In [29]:
df_victimas.isnull().sum()

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64

In [30]:
df_victimas.head()

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

procedemos a eliminar columnas rebundantes y transformar las columnas con nuestra funcion, ytransformar tipos de datos

In [31]:
columnas = ['AAAA', 'MM', 'DD']
df_victimas = df_victimas.drop(columnas, axis=1)

In [32]:
columnas = ['ROL', 'VICTIMA','SEXO']
transformar_dataframe(df_victimas, columnas)

In [33]:
renombrar = {'ID_hecho': 'Id_hechos',
             'FECHA': 'fecha',
             'ROL': 'rol',
             'VICTIMA': 'victima',
             'SEXO': 'sexo',
             'EDAD': 'edad',
             'FECHA_FALLECIMIENTO':'fecha_fallecimiento'}

df_victimas = df_victimas.rename(columns=renombrar)

In [36]:
df_victimas.sample(15)

Id_hechos      fecha                   rol victima       sexo edad  \
683  2021-0064 2021-08-02                Peaton  Peaton  Masculino   37   
170  2017-0032 2017-03-12  Pasajero_acompañante    Auto   Femenino   40   
407  2018-0117 2018-11-04             Conductor    Moto  Masculino   30   
414  2018-0124 2018-11-15             Conductor  Cargas  Masculino   44   
559  2020-0021 2020-03-14                Peaton  Peaton   Femenino   87   
307  2018-0021 2018-02-22                Peaton  Peaton   Femenino   62   
484  2019-0048 2019-06-07             Conductor    Moto  Masculino   32   
555  2020-0017 2020-03-05                Peaton  Peaton  Masculino   SD   
660  2021-0041 2021-05-20                Peaton  Peaton  Masculino   75   
98   2016-0126 2016-09-18             Conductor    Auto  Masculino   37   
435  2019-0001 2019-01-08             Conductor    Moto  Masculino   25   
332  2018-0044 2018-04-28                Peaton  Peaton   Femenino   50   
99   2016-0126 2016-09-18  Pasajero_acompañante    Auto  Masculino   60   
318  2018-0031 2018-03-28             Conductor    Moto  Masculino   31   
186  2017-0050 2017-04-28             Conductor    Moto  Masculino   46   

     fecha_fallecimiento  
683  2021-08-03 00:00:00  
170  2017-03-12 00:00:00  
407  2018-11-08 00:00:00  
414  2018-11-15 00:00:00  
559  2020-03-14 00:00:00  
307  2018-02-22 00:00:00  
484  2019-06-07 00:00:00  
555  2020-03-05 00:00:00  
660  2021-05-20 00:00:00  
98                    SD  
435  2019-01-08 00:00:00  
332  2018-04-28 00:00:00  
99                    SD  
318  2018-03-28 00:00:00  
186  2017-04-28 00:00:00

## MERGE

In [39]:
df = pd.merge(df_hechos, df_victimas, on='Id_hechos', how='inner')

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Id_hechos             717 non-null    object        
 1   nro_victimas          717 non-null    int64         
 2   fecha_x               717 non-null    datetime64[ns]
 3   hora                  717 non-null    object        
 4   lugar_hecho           717 non-null    object        
 5   tipo_calle            717 non-null    object        
 6   altura                133 non-null    float64       
 7   comuna                717 non-null    int64         
 8   geocodificacion_CABA  717 non-null    object        
 9   longitud              717 non-null    object        
 10  latitud               717 non-null    object        
 11  vehiculo_victima      717 non-null    object        
 12  vehiculo_acusado      717 non-null    object        
 13  fecha_y             

eliminamos una columna fecha ya que se instancia nuevamente

In [45]:
df.drop(columns=['fecha_y','victima'], inplace=True)
df.rename(columns={'fecha_x': 'fecha'}, inplace=True)

In [47]:
df.head()

Id_hechos  nro_victimas      fecha      hora  \
0  2016-0001             1 2016-01-01  04:00:00   
1  2016-0002             1 2016-01-02  01:15:00   
2  2016-0003             1 2016-01-03  07:00:00   
3  2016-0004             1 2016-01-10  00:00:00   
4  2016-0005             1 2016-01-21  05:20:00   

                                 lugar_hecho tipo_calle  altura  comuna  \
0  Av piedra buena y av fernandez de la cruz    Avenida     NaN       8   
1          Av gral paz  y av de los corrales   Gral paz     NaN       9   
2                         Av entre rios 2034    Avenida  2034.0       1   
3      Av larrazabal y gral villegas conrado    Avenida     NaN       8   
4   Av san juan y presidente luis saenz peña    Avenida     NaN       1   

                      geocodificacion_CABA      longitud       latitud  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

  vehiculo_victima vehiculo_acusado        rol       sexo edad  \
0             Moto             Auto  Conductor  Masculino   19   
1             Auto        Pasajeros  Conductor  Masculino   70   
2             Moto             Auto  Conductor  Masculino   30   
3             Moto               Sd  Conductor  Masculino   18   
4             Moto        Pasajeros  Conductor  Masculino   29   

   fecha_fallecimiento  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [48]:
duplicates = df[df.duplicated()]
duplicates

Empty DataFrame
Columns: [Id_hechos, nro_victimas, fecha, hora, lugar_hecho, tipo_calle, altura, comuna, geocodificacion_CABA, longitud, latitud, vehiculo_victima, vehiculo_acusado, rol, sexo, edad, fecha_fallecimiento]
Index: []

In [49]:
df.isnull().sum()

Id_hechos                 0
nro_victimas              0
fecha                     0
hora                      0
lugar_hecho               0
tipo_calle                0
altura                  584
comuna                    0
geocodificacion_CABA      0
longitud                  0
latitud                   0
vehiculo_victima          0
vehiculo_acusado          0
rol                       0
sexo                      0
edad                      0
fecha_fallecimiento       0
dtype: int64

In [50]:
# df.to_csv('Homicidios_hechos_victimas.csv',index=False)